core/src/main/java/io/grpc/internal/ForwardingClientStream.java

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright 2018, gRPC Authors All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.internal;
+
+import io.grpc.Attributes;
+import io.grpc.Compressor;
+import io.grpc.DecompressorRegistry;
+import io.grpc.Status;
+import java.io.InputStream;
+
+abstract class ForwardingClientStream implements ClientStream {
+  protected abstract ClientStream delegate();
```

Comment:
I mean tests to verify all methods are forwarded _reflectively_, to prevent bugs in the future when adding new methods to the base class, the forwarding classes are not forwarding them because people forget there are forwarding classes.

Reply:
I've been updating https://github.com/grpc/grpc-java/pull/3480 for a reflection unit test util. At the moment I think that PR is sufficient as a change detector, but longer term I think it would be nice to generate parameters reflectively and verify the forwarding class correctly forwards parameters, return values, and exceptions.